### PR TITLE
9523 - Pluggable Middleware

### PR DESCRIPTION
**Pull request by emsameen**
update sdk version

## Describe your changes

Implement pluggable middleware abstraction that supports Native or Dapr. The concrete implementation of the middleware includes:
### Service location 
- Dapr: via GRPC port
- Native: Env Variables 

### PubSub Client 
- Dapr: via Dapr pubSub 
- Native: Eclipse Paho MQTT client 

## Checklist - Manual tasks

* [x] Examples are executing successfully
* [x] Created/updated unit tests. Code Coverage percentage on new code shall be >= 80%.
* [x] Created/updated integration tests.
* [ ] Devcontainer can be opened successfully
* [x] Devcontainer can be opened successfully behind a corporate proxy
* [ ] Devcontainer can be re-built successfully
* [ ] Extended the documentation (e.g. README.md, CONTRIBUTING.md, Velocitas)

### CODE REVIEW

File: sdv/vdb/client.py [hunk 1]
```diff
@@ -41,16 +40,20 @@ class VehicleDataBrokerClient:
     def __new__(cls, port: Optional[int] = None):
```
Comment (BjoernAtBosch) [inline sdv/vdb/client.py:40]: Why is this `__new__` and not `__init__`?

File: sdv/vdb/client.py [hunk 2]
```diff
@@ -41,16 +40,20 @@ class VehicleDataBrokerClient:
...
-                    port = int(str(os.getenv("DAPR_GRPC_PORT")))
-                cls._address = f"localhost:{port}"
+            service_locator = config.middleware.service_locator
+            _location = service_locator.get_service_location("vehicledatabroker")
+            _hostname = urlparse(_location).hostname
+            if port is None:
+                _port = urlparse(_location).port
             else:
-                cls._address = conf.VEHICLE_DATA_BROKER_ADDRESS
+                _port = port
```
Comment (BjoernAtBosch) [inline sdv/vdb/client.py:49]: I seems a bit strange, that possible content of `port` parameter gets precedence over a possible port set in the `_location`.

File: sdv/vdb/client.py [hunk 3]
```diff
@@ -41,16 +40,20 @@ class VehicleDataBrokerClient:
...
+
+            _address = f"{_hostname}:{_port}"
```
Comment (BjoernAtBosch) [inline sdv/vdb/client.py:51]: Should be checked if `_port` is `None`, otherwise we end up with "<hostname>:" ...

File: sdv/model.py [hunk 4]
```diff
@@ -832,9 +832,13 @@ class Service(Node):
...
     def __init__(self):
         super().__init__()
-        _address = conf.service_locator.get_location(self.name)
-        self.channel = grpc.aio.insecure_channel(_address)  # type: ignore
-        self.metadata = conf.service_locator.get_metadata(self.name)
+        service_locator = config.middleware.service_locator
+        _location = service_locator.get_service_location(self.name)
+        _hostname = urlparse(_location).hostname
+        _port = urlparse(_location).port
+        self.address = f"{_hostname}:{_port}"
```
Comment (BjoernAtBosch) [inline sdv/model.py:839]: Should be checked if `_port` is `None`, otherwise we end up with "<hostname>:" ...

**Review by BjoernAtBosch — CHANGES_REQUESTED** [review]
I'd request some changes. Tests not yet looked at

File: sdv/base.py [hunk 5]
```diff
@@ -12,16 +12,67 @@
...
+    @abstractmethod
+    async def init(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def run(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def register_topic(self, topic: str, coro):
```
[the remark this answers is not in the document]
Comment (emsameen) [reply sdv/base.py:52]: done

File: .vscode/launch.json [hunk 6]
```diff
@@ -8,19 +8,47 @@
...
+            "program": "${workspaceFolder}/examples/seat-adjuster/src/main.py",
+            "console": "integratedTerminal",
+            "env": {
+                "SDV_MIDDLEWARE_TYPE": "native",
+                "SDV_MQTT_ADDRESS": "mqtt://localhost:1883",
+                "SDV_VEHICLEDATABROKER_ADDRESS": "grpc://localhost:55555",
+                "SDV_SEATSERVICE_ADDRESS": "grpc://localhost:50051"
+            }
+        },
+        {
```
[the remark this answers is not in the document]
Comment (emsameen) [reply .vscode/launch.json:39]: As discussed: descriptive label is added to every app launch

File: .vscode/scripts/run-feedercan.sh [hunk 7]
```diff
@@ -61,8 +61,16 @@ else
...
 
-dapr run \
+if [ $1 == "DAPR" ]; then
+  echo "Run Dapr ...!"
+  dapr run \
     --app-id feedercan \
     --app-protocol grpc \
     --components-path $ROOT_DIRECTORY/.dapr/components \
     --config $ROOT_DIRECTORY/.dapr/config.yaml -- python3 dbcfeeder.py
+else
```
[the remark this answers is not in the document]
Comment (emsameen) [reply .vscode/scripts/run-feedercan.sh:71]: done

File: .vscode/scripts/run-seatservice.sh [hunk 8]
```diff
@@ -50,11 +50,19 @@ export DAPR_GRPC_PORT=$SEATSERVICE_GRPC_PORT
...
+    --dapr-grpc-port $SEATSERVICE_GRPC_PORT \
+    --components-path $ROOT_DIRECTORY/.dapr/components \
+    --config $ROOT_DIRECTORY/.dapr/config.yaml & \
+    $SEATSERVICE_EXEC_PATH/val_start.sh
+else
+  echo "Run native ...!"
+  SEATSERVICE_GRPC_PORT='55555'
+  export DAPR_GRPC_PORT=$SEATSERVICE_GRPC_PORT
   $SEATSERVICE_EXEC_PATH/val_start.sh
+fi
```
[the remark this answers is not in the document]
Comment (emsameen) [reply .vscode/scripts/run-seatservice.sh:71]: Done

File: .vscode/scripts/run-seatservice.sh [hunk 9]
```diff
@@ -50,11 +50,19 @@ export DAPR_GRPC_PORT=$SEATSERVICE_GRPC_PORT
...
+  echo "Run Dapr ...!"
+  dapr run \
+    --app-id seatservice \
+    --app-protocol grpc \
+    --app-port $SEATSERVICE_PORT \
+    --dapr-grpc-port $SEATSERVICE_GRPC_PORT \
+    --components-path $ROOT_DIRECTORY/.dapr/components \
+    --config $ROOT_DIRECTORY/.dapr/config.yaml & \
+    $SEATSERVICE_EXEC_PATH/val_start.sh
+else
```
[the remark this answers is not in the document]
Comment (emsameen) [reply .vscode/scripts/run-seatservice.sh]: done

File: .vscode/scripts/run-vehicledatabroker.sh [hunk 10]
```diff
@@ -48,12 +48,17 @@ then
...
+  echo "Run Dapr ...!"
+  dapr run \
   --app-id vehicledatabroker \
   --app-protocol grpc \
   --app-port $DATABROKER_PORT \
   --dapr-grpc-port $DATABROKER_GRPC_PORT \
   --components-path $ROOT_DIRECTORY/.dapr/components \
   --config $ROOT_DIRECTORY/.dapr/config.yaml & \
   $DATABROKER_EXEC_PATH/databroker --address 0.0.0.0 --dummy-metadata
+else
```
[the remark this answers is not in the document]
Comment (emsameen) [reply .vscode/scripts/run-vehicledatabroker.sh]: done

File: .vscode/scripts/run-vehicledatabroker.sh [hunk 11]
```diff
@@ -48,12 +48,17 @@ then
...
+  echo "Run native ...!"
+  $DATABROKER_EXEC_PATH/databroker --address 0.0.0.0 --dummy-metadata
+fi
```
[the remark this answers is not in the document]
Comment (emsameen) [reply .vscode/scripts/run-vehicledatabroker.sh:67]: done

File: sdv/native/mqtt.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        self.callback = callback
+
+
+class MqttClient(PubSubClient):
+    """This class is a wrapper for the on_message callback of the MQTT broker."""
+
+    def __init__(self, port: Optional[int] = None, hostname: Optional[str] = None):
+        self._port = port
+        self._hostname = hostname
+        self._registered_topics: list[MqttTopicSubscription] = []
```
[the remark this answers is not in the document]
Comment (emsameen) [reply sdv/native/mqtt.py]: done

File: sdv/native/mqtt.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    async def run(self):
+        self._sub_client.loop_start()
+
+    async def init(self):
+        """Do nothing"""
+
+    async def register_topic(self, topic, coro):
+        if not self._sub_client.is_connected():
+            self._registered_topics.append(MqttTopicSubscription(topic, coro))
+        else:
```
[the remark this answers is not in the document]
Comment (emsameen) [reply sdv/native/mqtt.py]: done

File: sdv/vdb/client.py (hunk 1 above)
Comment (emsameen) [reply sdv/vdb/client.py:40]: This is used for the Singleton implementation of the VDBClient

File: sdv/vdb/client.py (hunk 2 above)
Comment (emsameen) [reply sdv/vdb/client.py:49]: The port can be overwritten by the developer and may replace the one from the service locator.

File: sdv/vdb/client.py (hunk 3 above)
Comment (emsameen) [reply sdv/vdb/client.py:51]: As discussed, unnecessary check. That will be added to the documentation for the service name convention.

File: sdv/model.py (hunk 4 above)
Comment (emsameen) [reply sdv/model.py:839]: related to the one above!

**Review by BjoernAtBosch — APPROVED** [review]
LGTM

**Pull request merged**